Add Tobira module for Synchronization with an external Tobira application

LukasKalbertodt opened the pull request:
This PR adds a new module `tobira`. This mainly offers the `/tobira/harvest` API that is used to synchronize Opencast data with an external Tobira. Using the `since` parameter, it is possible to get incremental updates, i.e. all data that changed since the given timestamp. Currently, the API only returns information about episodes and series.

There have been some previous PRs that changed existing code to meet the requirements of the Tobira module: #2762 #3075 #3076 #3119

For reviewers: I kept all original commits as I think they provide some useful context. However, it's probably most useful to just review the resulting code, instead of each commit individually. 


## FAQ

### Why merge Tobira-related code into upstream (this community repo)?

Unfortunately, maintaining and including external Opencast modules is a hassle and has a number of disadvantages. We think it is worth having the Tobira module in the community-Opencast to make sure it's always working and to reduce the hassle for adopters who want to use Tobira.

There is quite a bit of interest in Tobira from various universities (including ETH Zürich, Bern, TU Wien, Osnabrück, Siegen, and more). Especially compared with some other modules included in the community version, we think the Tobira module will be used quite a lot. It also makes it a lot easier for organizations to quickly test Tobira to evaluate whether they want to use it.

Finally, this Opencast module doesn't hurt any adopter not using it, except probably eating a tiny amount of RAM. It just offers an API, but otherwise is completely idle. Also, the API can only be used by users with admin privileges.


### Why merge this into `r/12.x` instead of `develop`?

Minor Opencast releases can include new features, but no breaking changes. This is very clearly just a new feature and doesn't break anything. It is practically useful as at least three universities are planning on using Tobira with Opencast 12. Thus, we can avoid some work there, if this already lands in `12.x`.

Also, as mentioned in the previous section, this module is completely passive and won't bother anyone not using it.


### How is the API versioned?

We have `/tobira/version` returning a `x.y` version number. That's just used for Tobira internally and doesn't mean the API may be used externally. The API's docs contain a note about that. For information on how the versioning works, see the comment in `TobiraApi.java`.


### Why does Tobira have to synchronize with Opencast this way?

One important requirement towards Tobira was that it still works even if Opencast is down (assuming static files are still served). Thus, Tobira needs a copy/cache of the relevant Opencast data. Additionally, if Tobira would always talk to Opencast APIs every time it needed some information, it would slow down Tobira considerably. 


### Why should this module run on the presentation node?

This module requests quite a bit of data from the search service, which runs on presentation nodes. To reduce overhead, the Tobira module should run on the same node.


---

### Your pull request should…

* [x] have a concise title
* [x] [close an accompanying issue](https://help.github.com/en/articles/closing-issues-using-keywords) if one exists
* [x] [be against the correct branch](https://docs.opencast.org/develop/developer/development-process#acceptance-criteria-for-patches-in-different-versions)
* [x] include migration scripts and documentation, if appropriate
* [x] pass automated tests
* [x] have a clean commit history
* [x] [have proper commit messages (title and body) for all commits](https://medium.com/@steveamaza/e028865e5791)

lkiesow committed:
Initial Tobira API Module

This patch adds a simple API module for Tobira which just returns static
JSON files for now.

LukasKalbertodt committed:
Obtain search service in `tobira`

Unfortunately, apparently you cannot mix `@Component` annotations with
modules that use the older XML based component definitions. Since we
need to use the search service (which uses the older XML definitions),
we need to use those as well. We think?

LukasKalbertodt committed:
Add `limit` and `since` parameter to Tobira API

LukasKalbertodt committed:
Implement first version of Tobira harvesting API that returns events

This is definitely not complete yet, but should work fine for events.

LukasKalbertodt committed:
Add series to Tobira harvesting API and generally improve the API

This commit unfortunately contains quite a few change, but it was
easier to fix everything in one go.

- The `includesItemsUntil` now is at least `TIME_BUFFER_SIZE` far in
  the past.
- Series are loaded and merged with events.
- Events and series are now loaded without checking ACLs and requiring
  an administrative context.

LukasKalbertodt committed:
Optimize `includesItemsUntil` in Tobira harvest API

This new logic is more precise and leads to fewer duplicate items.

LukasKalbertodt committed:
Rename `limit` to `preferredAmount` in Tobira harvest API

LukasKalbertodt committed:
Add `thumbnail` field to events in Tobira harvest response

LukasKalbertodt committed:
Add lots more data to events in the Tobira harvest API

LukasKalbertodt committed:
Use `Jsons.NULL` instead of `null` for missing resolution

Previously, this would lead to a null exception. The reason we can pass
`null` for many other fields is because there, the `Jsons.p` overload
for strings or whatever is used, which accepts `null`. But the `Jsons.p`
overload for `Jsons.Val` does not.

LukasKalbertodt committed:
Include ACLs for events in Tobira harvest API response

LukasKalbertodt committed:
Make duration in Tobira API non-null

Having this field nullable ripples through the whole Tobira codebase
and adds lots of special cases everywhere. We decided to ignore this
special case and just use a dummy duration of 0 instead. I know, I know,
this is bad of course, in some sense. But we can simply consider it a
bug if an event does not have a duration. Opencast should know and can
know. In our large test data set, none of the events have an unknown
duration. And what's the alternative? Having it nullable everywhere,
in Tobira's frontend, backend and DB. This adds a lot of code for
nothing basically. And if the type system doesn't catch it, it can
easily introduce weird bugs. So we decided to accept this slight hack
in order to simplify other parts. Worst case: Tobira shows a 0 duration
for some events.

LukasKalbertodt committed:
Fix duration in Tobira API

I am a dumbo

LukasKalbertodt committed:
Split Tobira module into three files

The main file has gotten fairly large, and this split somewhat makes
sense.

LukasKalbertodt committed:
Fix warning in Tobira code

LukasKalbertodt committed:
Explicitly mark "TODO" comment

LukasKalbertodt committed:
Use local variable type inference across Tobira code

We wanted to use that from the beginning, but it didn't easily work
back then. Now we finally can.

LukasKalbertodt committed:
Add `isLive` field for events in harvesting API

LukasKalbertodt committed:
Return full array of creators instead of single one in harvest API

This was an oversight. We should have returned the full list of creators
from the beginning. One could argue the same for other DC fields, like
title and description. However, multiple values here are much less
used and Tobira would not know how to interpret those.

LukasKalbertodt committed:
Apply code review suggestion

Co-authored-by: Julian Kniephoff <me@juliankniephoff.com>

LukasKalbertodt committed:
Include all additional event metadata in harvest response

LukasKalbertodt committed:
Get title from mediapackage instead of `getDcTitle`

The old method has problems, I was told, and is maybe outdated or weird.

LukasKalbertodt committed:
Extract more code into separate methods

LukasKalbertodt committed:
Add `timelinePreview` object to Tobira harvest API

LukasKalbertodt committed:
Include all catalogs for event metadata in Tobira harvest API

LukasKalbertodt committed:
Make catalog filtering code more defensive to deal with `null` better

A null mime type caused some problems.

LukasKalbertodt committed:
Add Tobira module to all presentation node assemblies

Tobira should run on the presentation node to avoid overhead of the
search service remote impl.

LukasKalbertodt committed:
Add `/tobira/version` endpoint

LukasKalbertodt committed:
Improve exception logging from `new Item` in harvest API

LukasKalbertodt committed:
Add `acl` field to series items (like events already had)

LukasKalbertodt committed:
Remove unneeded specification of the content type in Tobira API

LukasKalbertodt committed:
Skip broken items instead of returning 500

Before this, any exception in `Item` constructors (where we retrieve the
information about events/series) would lead to us returning 500 from
the Tobira API. Over time we noticed, that these exception sadly happen
from time to time because events or series are just broken. And don't
have data we assume they have.

While this runs the risk of administrators never noticing that an event
is broken and thus not synchronized with Tobira, we just log an error
and skip the item. We don't want to outright make synchronization
impossible when there is a broken item. This should be combined with
proper monitoring to notice whens something like this goes wrong.

LukasKalbertodt committed:
Add `startTime` and `endTime` to the Tobira harvest response for events

LukasKalbertodt committed:
Try all three sources to get an event title for Tobira API

This should also reduce the number of times we load the DCC catalogs.

LukasKalbertodt committed:
Bump version of Tobira module to 12

LukasKalbertodt committed:
Use OSGI annotations instead of XML file in Tobira module

LukasKalbertodt committed:
Remove unused resource in Tobira module

LukasKalbertodt committed:
Add general Tobira API documentation

geichelberger commented on modules/tobira/src/main/java/org/opencastproject/tobira/impl/TobiraApi.java:
```diff
@@ -0,0 +1,207 @@
+    immediate = true,
+    service = TobiraApi.class
+)
+public class TobiraApi {
```
I would rename the class to match the other "REST"endpoints. e.g. TobiraEndpoint

geichelberger commented on modules/tobira/pom.xml:
```diff
@@ -0,0 +1,92 @@
+            <Export-Package>
+              org.opencastproject.tobira;version=${project.version}
+            </Export-Package>
+            <!-- Remove once all used services use OSGi annotations -->
+            <_dsannotations-options>norequirements</_dsannotations-options>
```
It should be possible to remove this.

LukasKalbertodt replied:
Oh true, somehow I thought it was necessary :shrug:

geichelberger reviewed (changes requested):
An `intercept-url` entry is missing for the Tobira endpoint in the security.xml.

LukasKalbertodt committed:
Add entries Tobira paths in security.xml

The harvest endpoint can currently only used by administrators anyway.
And before this commit, `/tobira/**` was also only accessible to admins.
This allows anyone to access `/tobira/version` as this just shows the
version of the module, i.e. information that does not need to be
protected.

LukasKalbertodt committed:
Rename class `TobiraApi` to `TobiraEndpoint`

This brings it more in line with other classes providing "REST"
endpoints. The `*Api` suffix is usually used differently in Opencast.

LukasKalbertodt committed:
Remove superfluous entry in pom.xml

geichelberger reviewed (approved):
LGTM

It feels like every bundle uses a different way to build JSON Responses ;)
